### ISSUE TITLE
Subclasses in dict values not inferred as subclasses

### ISSUE THREAD

**Issue by adamjstewart** [astral-sh/ty#2985]
### Summary

The following code:
```python
from torchmetrics import MeanAbsoluteError, MeanSquaredError, MetricCollection

MetricCollection({
    'mse': MeanSquaredError(),
    'mae': MeanAbsoluteError(),
})
```
fails in ty 0.0.21 (but not prior versions):
```console
> ty check test.py
error[invalid-argument-type]: Argument to bound method `__init__` is incorrect
 --> test.py:3:18
  |
1 |   from torchmetrics import MeanAbsoluteError, MeanSquaredError, MetricCollection
2 |
3 |   MetricCollection({
  |  __________________^
4 | |     'mse': MeanSquaredError(),
5 | |     'mae': MeanAbsoluteError(),
6 | | })
  | |_^ Expected `Metric | MetricCollection | Sequence[Metric | MetricCollection] | dict[str, Metric | MetricCollection]`, found `dict[str, MeanSquaredError | MeanAbsoluteError]`
  |
info: Method defined here
   --> /Users/Adam/spack/var/spack/environments/default/.spack-env/._view/gd2vtbwe4fojggua4emapinyxnqhryg4/lib/python3.14/site-packages/torchmetrics/collections.py:199:9
    |
197 |       __jit_unused_properties__: ClassVar[list[str]] = ["metric_state"]
198 |
199 |       def __init__(
    |           ^^^^^^^^
200 |           self,
201 | /         metrics: Union[
202 | |             Metric,
203 | |             "MetricCollection",
204 | |             Sequence[Union[Metric, "MetricCollection"]],
205 | |             dict[str, Union[Metric, "MetricCollection"]],
206 | |         ],
    | |_________- Parameter declared here
207 |           *additional_metrics: Metric,
208 |           prefix: Optional[str] = None,
    |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```
However, `MeanSquaredError` and `MeanAbsoluteError` are subclasses of `Metric`, so I believe this should be valid. For now, easiest workaround is to store these in a variable with explicit type hints.

### Version

ty 0.0.21

**Comment by carljm** [comment]
Thanks for the report! Yes, this is because we stopped unioning `Unknown` into the value types of un-annotated container literals (discussed in #1240); the `Unknown` previously allowed this to pass but also allowed many false negatives.

It looks like we need to improve our bidirectional inference here, particularly in the case where there's a union with ~~a generic protocol~~ ...something? Here's a self-contained standalone repro:

```py
from typing import Sequence

class Metric: ...

class M1(Metric): ...
class M2(Metric): ...

def f(d: dict[str, Metric] | Sequence[Metric]): ...

f({"one": M1(), "two": M2()})
```

If we replace `Sequence[Metric]` here with `list[Metric]`, or with `Metric`, our bidirectional inference succeeds and the error goes away.

**Comment by AlexWaygood** [comment]
> particularly in the case where there's a union with a generic protocol.

hmm, `Sequence` isn't a protocol, it's just a nominal class

**Comment by carljm** [comment]
I wonder if the issue is related to the fact that a `dict` is also a `Sequence`.

**Comment by AlexWaygood** [comment]
> I wonder if the issue is related to the fact that a `dict` is also a `Sequence`.

er, no it isn't 😅 https://play.ty.dev/027ca911-a5fd-43b3-99cb-16a4c8d04f30

**Comment by carljm** [comment]
Ok, scratch that theory :) I guess it's iterable but not a full sequence.

**Comment by carljm** [comment]
It's because `dict` and `Sequence` are not disjoint. (But `dict` and `list` are.) We filter provably disjoint union elements, and if more than one union element remains, we currently give up. Full fix would be multi inference with remaining elements; experimenting with an alternative where we prefer union elements with a subtype relationship (which would filter out Sequence here).

**Comment by carljm** [comment]
This is fixed by https://github.com/astral-sh/ruff/pull/23844, which goes ahead and does full multi-inference for union type contexts.
